### PR TITLE
Implement Actions and cosmetics and minor fixes

### PR DESCRIPTION
**Pull request by BigRoy**
This is becoming quite a big PR so I'm setting this up now so we can discuss the current differences. I'm looking to work on embedding a dark style sheet soon too, which will contain some larger files of its own and I'd rather have that as a PR by itself.

> With the actions this might require the updated launcher, but I'm expecting this actually also works with the older launcher. (Though then it just doesn't use the actions)

This PR currently implements:

- Actions (and Application subclass) that will be available to the [new Launcher update.](https://github.com/Colorbleed/launcher/commit/87dc3e72a8d9cc116074685199a21502ce334cef) The update will also need a PR of its own.
- This improves `get_application` with Python 2 encoded environment variables, a fix ported over from Launcher.
- This ports over `launch` from Launcher so it's in the core.
- This adds the current context to the top of the Avalon menu in Maya - where one can also "set the context" with the set context tool.

avalon_menu_context

- Implements `__main__` for `avalon.tools.cbloader` with a command line interface (similar to projectmanager)
- Fix `cbloader` pretty date delegate so that it uses `datetime.now()` similar to how avalon api exposes `api.time()`
- Fix cbloader pretty delegate by forcing integer division for *python 3 compatibility*
- [Fix exceptions for *Python 3 compatibility*](https://github.com/getavalon/core/compare/master...Colorbleed:master#diff-bd79ed282e6d11cd8ded6dc7bdb10acaR81)
- Cosmetics: [visually remove superfluous zeros in cbloader from frame ranges](https://github.com/getavalon/core/compare/master...Colorbleed:master#diff-2aba52af6f402a3f3e5110176faec69dR66)
- Add family filter in `cbloader` (see screenshot below)
- Add support for family _icons_ and _label_ in `cbloader` in the project's configuration:

```python
"config" {
    "families" : [ 
                {
                    "name" : "colorbleed.animation",
                    "icon" : "male",
                    "label" : "Animation"
                }, 
                {
                    "name" : "colorbleed.camera",
                    "icon" : "video-camera",
                    "state": false,
                    "label" : "Camera"
                },
    ]
}
```

loader_family_icons

> Note the above code sample (above screenshots) shows the camera family configuration having a state of `false` (this is a temporary support for default states of the filters in the cbloader) I'm expecting to move this to something more user-oriented where the user can pick their defaults.

- Added support for task icons in project config for the task model in the project manager (and the context manager)

```python
"config" : {
        "tasks" : [ 
            {
                "name" : "model",
                "icon" : "user-plus"
            }, 
            {
                "name" : "render",
                "icon" : "lightbulb-o"
            }
    ]
}
```

- Slight cosmetic changes to the asset widget used in `cbloader` and `projectmanager`; the filter view is now at the top.

### CODE REVIEW

**Comment by mottosso** [comment]
A bit large yes, but thanks for putting together that summary. Makes it easier to grok.

> Note the above camera **family** configuration having a state of false

Did you mean the Look family?

**Comment by BigRoy** [comment]
> Did you mean the Look family?

A good, point. I can see the screenshot is confusing since it has a different state. I was referring to the code sample above it. :)

**Comment by mottosso** [comment]
Ah, yes I wasn't sure at first.

About `"state": false`, this to me sounds like state has been turned off, but what state could it hold? Whether it is checked or not I think is part of it's state, and with this it reads (to me) as though it wouldn't be able to store any state.

How about calling it `"active": false"`?

**Comment by BigRoy** [comment]
> How about calling it "active": false"?

Could work. It'll still be active. It's a temporary workaround though, until we somehow store user preferences with the tools (like their own defaults). 

Maybe something more explicit like `"defaultToActive": false` makes more sense?

**Comment by BigRoy** [comment]
Also, here's some more info about the `Actions`.

# Actions

The actions are registered and discovered in the same way as `Loader` and `Creator` plug-ins but inherit from `api.Actions`.

An implemented for an Action can look like:

```python
import avalon.api
from avalon import lib


class ProjectManagerAction(avalon.api.Action):
    name = "projectmanager"
    label = "Project Manager"
    icon = "gear"
    order = 999     # at the end

    def is_compatible(self, session):
        return "AVALON_PROJECT" in session

    def process(self, session, **kwargs):

        return lib.launch(executable="python",
                          args=["-u", "-m", "avalon.tools.projectmanager",
                                session['AVALON_PROJECT']])
```

To register it explicitly:
```python
avalon.api.register_plugin(avalon.api.Action, ProjectManagerAction)
```

With it registered you can find it back through the api:

```python
for action in api.discover(api.Action):
    print(action)
```
> Note that the `api.discover` method returns the class and *not* an instance of the class.

As such, to test whether an action can be triggered in the current context and if so to run it you would do:

```python
# Note: this would trigger *all* compatible actions within the current session
for Action in api.discover(api.Action):
    action = Action()
    session = api.Session.copy()  # currently active session
    if action.is_compatible(session):
        action.process(session)
```

**Comment by BigRoy** [comment]
@mottosso The travis check is failing on python files from the stylesheet now I merged it with the currently live master. It should just ignore those specific files because they should actually only be loaded whenever those Qt versions are actually live. Any idea how I can make it pass the travis builds?

Other than that, I think this is ready for a merge. We've been using the Actions/Applications as a basis for quite some time now and it's been going pretty well. With this merged I'd be happy to continue some more development and hopefully work towards smaller commits.

**Comment by mottosso** [comment]
Having a look

**Comment by mottosso** [comment]
Made a PR for you.

About the `AVALON_PROJECT` warnings, any idea what that's about? :O

File: run_maya_tests.py [hunk 1]
```diff
@@ -47,6 +47,9 @@
         # We can expect any vendors to
         # be well tested beforehand.
         "--exclude-dir=avalon/vendor",
+
```
[the remark this answers is not in the document]
Comment (mottosso) [reply run_maya_tests.py]: Ah, sorry about that, would you mind squashing this one?

**Comment by BigRoy** [comment]
> About the AVALON_PROJECT warnings, any idea what that's about? :O

Actually, no idea. It's actually mentioning `AVALON_PROJECTS` which is required for the schema and the setup in the test is not setting that, I could do so - that would avoid these warnings.

Shall I do that for now?

**Comment by mottosso** [comment]
I think so, yes. Hard to read the test results at the moment I think.

**Comment by BigRoy** [comment]
Actually, looking at the schema requirements:
```
     u'required': [u'AVALON_PROJECTS',
                   u'AVALON_PROJECT',
                   u'AVALON_ASSET',
                   u'AVALON_SILO',
                   u'AVALON_CONFIG'],
```
Then after this it will also log the warning for asset, silo and config. We might need to set a placeholder for all of those then.

**Comment by BigRoy** [comment]
> Let me test this PR tomorrow by the way, I think it's failing on some parts with these latest commits.

Ok, [fixed two issues from refactoring conflicts](https://github.com/getavalon/core/pull/306/commits/874aa77b0b28984233c6298d6c80d530bc887b72). :)
Also just committed a ["fix" for the required keys in tests where I'm assigning them placeholder values](https://github.com/getavalon/core/pull/306/commits/afc532b32881aa4579dceb61c53b6b5f4bd05774).

**Comment by mottosso** [comment]
Oo, very nice :)

**Comment by BigRoy** [comment]
> Let's see if the tests pass without warnings now.

Passed without the warning messages - indeed much more readable. :+1:

**Comment by BigRoy** [comment]
To me this is ready for a merge - who's going to click the big button? :smile:

**Comment by mottosso** [comment]
Go ahead, you've earned it! :)

**Pull request merged**